### PR TITLE
feat: live events

### PR DESCRIPTION
**Pull request by yakkomajuri**
Here's a WIP for live events with a bit of duct tape to make it work. 

It is not ready!! I'm putting this here for visibility on what I've been working on and to set the stage for sharing context.

A lot of things are missing, such as:

- [x] ~~Better design~~
- [x] Fix `ts` vs `timestamp` parsing issues in the serializers
- [x] Update plugin server handling of `events_plugin_ingestion` topic messages
- [x] Feature flagging this
- [x] Testing
- [x] Code quality

Here's a demo though:

https://user-images.githubusercontent.com/38760734/204628649-373ab9f8-6f35-437b-8556-4ef529f3156e.mov

### CODE REVIEW

File: posthog/models/live_events/sql.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+LIVE_EVENTS_MV_TABLE_SQL = lambda: """
+CREATE MATERIALIZED VIEW live_events_mv ON CLUSTER '{cluster}'
+TO {database}.{target_table}
+AS SELECT
+	uuid,
+	event,
+    properties,
+	-- KLUDGE: we don't have the real timestamp yet
```
Comment (yakkomajuri) [inline posthog/models/live_events/sql.py:67]: porting `handleTimestamp` to the Django side is the best solution here and will also help with a potential fix for historical imports

**Comment by yakkomajuri** [comment]
I've decided to punt on design based on our handbook:

https://posthog.com/handbook/engineering/product-design#v01-or-v2

It seems like a simple thing, but actually there's one little hard-ish problem which is that we currently only poll for new events, so old events that are marked as "being processed" will only lose that tag if you refresh the page. Otherwise there's quite a bit of a refactor to be done. 

Nevertheless, I've looped in Cory here #13042 so we can continue the discussion there once we're ready to remove the feature flag.

File: posthog/api/test/test_event.py [hunk 2]
```diff
@@ -686,3 +686,694 @@ def test_filter_events_with_date_format(self):
 
         self.assertEqual(len(response["results"]), 1)
         self.assertEqual([r["event"] for r in response["results"]], ["should_be_included"])
+
+
+# Copy paste of TestEvents testing everything with the live-events flag enabled
+# Eventually we'll remove the above class and only use this one once the flag is
+# gone and we enable the feature for everyone
+class TestLiveEvents(ClickhouseTestMixin, APIBaseTest):
```
Comment (yakkomajuri) [inline posthog/api/test/test_event.py:694]: this shouldn't need a review as it's literally copy paste from the class above, just that all methods got this decorator:

```
@patch("posthoganalytics.feature_enabled", return_value=True)
```

There's a way to do this by creating a decorator that we'd add to all methods on the class above but this is the easiest way and actually cleanest / easiest to debug. These tests helped me find a bunch of bugs!

File: posthog/models/live_events/sql.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        uuid,
+        event,
+        team_id,
+        distinct_id,
+        argMax(properties, created_at) as properties_aggregated,
+        argMax(elements_chain, created_at) as elements_chain_aggregated,
+        argMax(timestamp, created_at) as timestamp_aggregated,
+        argMax(_table, created_at) as source_table
+    FROM
+        merge('{database}', '^(events|live_events)$')
```
Comment (hazzadous) [inline posthog/models/live_events/sql.py:106]: Neat!

I don't see this referenced in any of the snapshots, is this being tested?

File: posthog/api/capture.py [hunk 4]
```diff
@@ -50,6 +50,10 @@ def parse_kafka_event_data(
     token: str,
 ) -> Dict:
     logger.debug("parse_kafka_event_data", token=token, team_id=team_id)
+    properties = json.dumps(data["properties"])
```
Comment (hazzadous) [inline posthog/api/capture.py:53]: Is it possible that "properties" isn't in the data dict? Will we raise a 400?

File: plugin-server/tests/utils/event.test.ts [hunk 5]
```diff
@@ -134,7 +134,47 @@ describe('formPipelineEvent()', () => {
             $set_once: {},
         })
     })
+    it('forms pluginEvent from a raw message with properties and event at the top level', () => {
```
Comment (hazzadous) [inline plugin-server/tests/utils/event.test.ts:137]: Should we update the tests to use the new structure? And rather add a tests for backwards compatible structure?

File: posthog/api/test/test_event.py [hunk 6]
```diff
@@ -686,3 +686,694 @@ def test_filter_events_with_date_format(self):
...
+                distinct_id="bla",
+                event=event,
+                team=self.team,
+                properties={"random_prop": "don't include", "some other prop": "with some text"},
+            )
+        response = self.client.get(f"/api/projects/{self.team.id}/events/values/?key=custom_event").json()
+        self.assertListEqual(sorted(events), sorted(event["name"] for event in response))
+
+    @patch("posthoganalytics.feature_enabled", return_value=True)
+    @test_with_materialized_columns(["random_prop"])
```
Comment (hazzadous) [inline posthog/api/test/test_event.py:846]: How do materialized columns work with the live events table? Presumable the events and live_events tables need to be in sync to be able to generate a valid query?

File: posthog/api/capture.py (hunk 4 above)
Comment (yakkomajuri) [reply posthog/api/capture.py:53]: it shouldn't be - although indeed looking at it in isolation like this does seem to show that. I can add some handling here but we do earlier on set it to `{}` if it's empty. Tests actually verify this.

File: posthog/models/live_events/sql.py (hunk 3 above)
Comment (yakkomajuri) [reply posthog/models/live_events/sql.py:106]: Yup, added those under `test_live_events`

File: plugin-server/tests/utils/event.test.ts (hunk 5 above)
Comment (yakkomajuri) [reply plugin-server/tests/utils/event.test.ts]: testing both cases extensively now

**Comment by yakkomajuri** [comment]
This behemoth is actually mergeable now!

**Review by hazzadous — APPROVED** [review]

**Comment by yakkomajuri** [comment]
damn - have to go back and add static uuids

**Comment by hazzadous** [comment]
Closing in favour of https://github.com/PostHog/posthog/pull/13191 which avoids some of the complexity added here.

**Pull request closed without merging**